### PR TITLE
server: allow resource reuse

### PR DESCRIPTION
**Pull request by praiskup**

### CODE REVIEW

File: bin/resalloc [hunk 1]
```diff
@@ -66,7 +76,7 @@ def main():
 
         conn = Connection(arg.connection)
         if 'ticket' == arg.subparser:
-            ticket = conn.newTicket(arg.tags)
+            ticket = conn.newTicket(arg.tags, sandbox=arg.sandbox)
```
Comment (phracek) [inline bin/resalloc:79]:
```suggestion
            ticket = conn.newTicket(tags=arg.tags, sandbox=arg.sandbox)
```

File: resalloc/client.py [hunk 2]
```diff
@@ -38,12 +38,16 @@ class Connection(object):
...
         if tags == None:
             raise ResallocClientException("no tags specified")
 
         t = Ticket(connection=self.connection)
-        t.id = self.connection.takeTicket(tags)
+        args = [tags]
+        if sandbox:
+            args.append(sandbox)
+
+        t.id = self.connection.takeTicket(*args)
```
Comment (phracek) [inline resalloc/client.py:50]:
```suggestion
        t.id = self.connection.takeTicket(tags=tags, sandbox=sandbox)
```

Condition is not needed. takeTicket by default gets None.

File: resallocserver/manager.py [hunk 3]
```diff
@@ -26,7 +26,8 @@
 from resallocserver import models
 from resallocserver.db import session_scope
 from resallocserver.log import get_logger
-from resallocserver.logic import QResources, QTickets
+from resallocserver.logic import (
+        QResources, QTickets, assign_ticket, release_resource)
```
Comment (phracek) [inline resallocserver/manager.py:30]:
```suggestion
        QResources, QTickets, assign_ticket, release_resource
)
```

File: resallocserver/logic.py [hunk 4]
```diff
@@ -41,34 +66,55 @@ def on(self):
...
         return self.query.filter_by(state=RState.STARTING)
 
     def stats(self):
         items = {}
-        items['on']     = self.on().outerjoin(models.Ticket).all()
+        items['on']     = self.on().all()
         items['up']     = [x for x in items['on'] if x.state  == RState.UP]
-        items['ready']  = [x for x in items['up'] if x.ticket == None]
-        items['taken']  = [x for x in items['up'] if x.ticket != None]
+        items['ready']  = [x for x in items['up'] if x.ticket_id is None]
```
Comment (phracek) [inline resallocserver/logic.py:92]:
```suggestion
        items['ready']  = [x for x in items['up'] if not x.ticket_id]
```

File: resallocserver/logic.py [hunk 5]
```diff
@@ -41,34 +66,55 @@ def on(self):
...
+        items['taken']  = [x for x in items['up'] if x.ticket_id is not None]
```
Comment (phracek) [inline resallocserver/logic.py:93]:
```suggestion
        items['taken']  = [x for x in items['up'] if x.ticket_id]
```

File: bin/resalloc (hunk 1 above)
Comment (praiskup) [reply bin/resalloc:79]: I don't want to change calling convention if not necessary.

File: resalloc/client.py (hunk 2 above)
Comment (praiskup) [reply resalloc/client.py:50]: This goes through xmlrpc, and `sandbox=` named argument doesn't work there.  If you have an idea how to make this work, let me know.

**Comment by praiskup** [comment]
thanks for the review!

File: resallocserver/logic.py (hunk 5 above)
Comment (praiskup) [reply resallocserver/logic.py:93]: I'm not sure it is safe to expect ticket != 0 .. please keep 'is None' here to only calculate NULL values

**Comment by praiskup** [comment]
ptal

File: resallocserver/logic.py (hunk 5 above)
Comment (phracek) [reply resallocserver/logic.py:93]: The original comprehensive list means that all items['taken'] values will be all except None. Everythong else is count.

File: bin/resalloc (hunk 1 above)
Comment (phracek) [reply bin/resalloc:79]: From python point of view, it does not look good, but it is up on you. `newTicket` used named variables and it should work.

File: resalloc/client.py (hunk 2 above)
Comment (phracek) [reply resalloc/client.py:51]: I am not able to find out `takeTicket` function in xmlrpc.client.ServerProxy https://github.com/python/cpython/blob/3.8/Lib/xmlrpc/client.py#L1389

Comment (phracek) [reply resalloc/client.py:51]: Let's leave it as you propose.

File: resallocserver/logic.py (hunk 5 above)
Comment (praiskup) [reply resallocserver/logic.py:93]: The semantics is `what resource is without assigned ticket` resp. `with assigned ticket`.  The ticket_id can probably be zero in rare cases.  I mean this:
```
In [1]: [x for x in [None, 0, "", 1] if x]                                                                                                                                                                                                   
Out[1]: [1]

In [2]: [x for x in [None, 0, "", 1] if x is not None]                                                                                                                                                                                       
Out[2]: [0, '', 1]
```

File: resalloc/client.py (hunk 2 above)
Comment (praiskup) [reply resalloc/client.py:51]: It is xmlrpc call to ServerAPI.

File: resallocserver/logic.py (hunk 5 above)
Comment (phracek) [reply resallocserver/logic.py:93]: I got it. Make sense.

File: bin/resalloc (hunk 1 above)
Comment (praiskup) [reply bin/resalloc]: I changed it so the call is consistent at least.   The `python3-resalloc` API would deserve some
orthogonal cleanup work, documentation, etc.  Let's postpone the work on this for now.

**Pull request merged**